Use grad_out for cudnn CTC loss

t-vi opened the pull request:
Using grad_out for CuDNN CTC loss fixes: #26797, #25833.

We also fix a cudnn incompatible change that surfaced during the testing: As of CuDNN 7.6 the semantics of the CTC loss gradients are different.
This leads us to disable CuDNN CTC for CuDNN < 7.6. To mitigate the impact on users, we convert the parameters for the native implementation if CuDNN isn't applicable (previously this would give an error.)

t-vi committed:
Use grad_out for cudnn CTC loss

Fixes: #26797, #25833.

t-vi committed:
tighten skip condition

t-vi commented:
The build failures are exactly my new test.
So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.
I cannot reproduce the failure locally with cuda10.1, I'll try cuda 9 next.

If you had a hint in addition, I'd also be very grateful.

soumith commented:
> So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.

@ezyang ideas?

ezyang commented:
> So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.

That's not my interpretation of the test failures. I see:

```
Sep 29 22:46:00 ======================================================================
Sep 29 22:46:00 FAIL: test_ctc_loss_cudnn_cpu (__main__.TestAutogradDeviceTypeCPU)
Sep 29 22:46:00 ----------------------------------------------------------------------
Sep 29 22:46:00 Traceback (most recent call last):
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 161, in instantiated_test
Sep 29 22:46:00     return test(self, device_arg)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 298, in dep_fn
Sep 29 22:46:00     return fn(slf, device, *args, **kwargs)
Sep 29 22:46:00   File "test_autograd.py", line 3856, in test_ctc_loss_cudnn
Sep 29 22:46:00     self.assertEqual(grad_cudnn, grad_native, prec=1e-4)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_utils.py", line 698, in assertEqual
Sep 29 22:46:00     assertTensorsEqual(x, y)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_utils.py", line 668, in assertTensorsEqual
Sep 29 22:46:00     self.assertLessEqual(max_err, prec, message)
Sep 29 22:46:00 AssertionError: tensor(836.3644, device='cuda:0', dtype=torch.float32) not less than or equal to 0.0001 : 
```

If the derivatives.yaml change were not picked up, I'd expect to see some sort of shape error (due the unsqueezes not being applied). My guess is that it's cuDNN version related.

t-vi commented:
Well I rebuilt locally with cuda 9 (but still cudnn 7.1.4) and it works. As far as I can tell it's cudnn 7.6 for the CI.  I will try upgrading cudnn here.

I'll try adding a backward to templates/Functions.cpp just to be sure (probably nicer, anyway, to not compute the gradient over and over in the ternary).

t-vi committed:
move ctc backward to templates/Functions.cpp

t-vi commented:
So @ezyang was right: With CuDNN 7.1, the patch works well, with CuDNN 7.6, I get the same failure as the CI.

ptrblck commented:
We are looking into it, as the test works fine in cudnn7500 and fails in cudnn7604.

t-vi committed:
Add version check on CuDNN

Thanks to @ptrblck for the advice.
All errors are mine. I must admit this is programmed against the CI as
I don't have hardware ready at the moment.

t-vi committed:
tab police

t-vi committed:
dereference

t-vi committed:
Merge branch 'master' into cudnn_ctc_gradout

t-vi commented:
Unfortunately, I think this is still very broken.

ngimel commented on aten/src/ATen/native/cudnn/LossCTC.cpp:
```diff
@@ -62,7 +62,13 @@ std::tuple<Tensor, Tensor> _cudnn_ctc_loss(const Tensor& log_probs_t, const Tens
 
-  Tensor probs = log_probs->softmax(2);
+  Tensor probs;
+  // before 7.6 passing probs was the only option, starting with 7.1 the log probs/unnormalized log probs is the default
```
starting with 7.6? Nice find, it's not in the release notes unfortunately

t-vi replied:
Unfortunately, the fix isn't working, either, as one would need to set the non-default option or somesuch.

ngimel commented on test/test_autograd.py:
```diff
@@ -3832,6 +3832,29 @@ def ctc_after_softmax(x):
             gradcheck(ctc_after_softmax, [x], nondet_tol=1e-7)
 
+    @skipCUDAIfRocm
+    @unittest.skipIf(not (TEST_CUDNN and (TEST_CUDNN_VERSION if TEST_CUDNN_VERSION else 0) >= 7000), "needs cudnn >= 7.0")
```
consider using device-generic cudnn decorators (added by #26791)

ngimel commented on test/test_autograd.py:
```diff
@@ -3832,6 +3832,29 @@ def ctc_after_softmax(x):
+        grad_out = torch.randn(batch_size, device='cuda', dtype=torch.float)
+        loss_native = torch.nn.functional.ctc_loss(log_probs, targets, input_lengths, target_lengths, reduction='none')
+        grad_native, = torch.autograd.grad(loss_native, log_probs, grad_out)
+        loss_cudnn = torch.nn.functional.ctc_loss(log_probs, targets.to('cpu', torch.int32),
```
As I see, the thing that triggers cudnn use (compared to line 3850) is targets being on the cpu and of int32 type? That seems fragile and non-obvious in the test, can you put line 3850 in the context manager disabling cudnn?

ngimel reviewed (approved):
approved, subject to minor fixes in the test

t-vi committed:
improve tests

t-vi committed:
gross hack to get the right gradient...

t-vi commented:
So this is all pretty terrible. As an alternative approach, I tried to use the default input options for CuDNN 7.6 and using softmax outside the "core function" (so that the backward of softmax is used), I do get differences in the derivatives. These are smaller and only for some inputs, apparently, but given that the native function does seem to pass gradcheck in the snippet below.

```python
batch_size = 1
input_length = 10
num_labels = 3
target_length = 5
targets = torch.randint(1, num_labels, (batch_size * target_length,),
                        device='cuda', dtype=torch.long)
log_probs = torch.log_softmax(torch.randn(input_length, batch_size, num_labels, device='cuda', dtype=torch.float), 2)
log_probs.requires_grad_()

input_lengths = batch_size * [input_length]
target_lengths = batch_size * [target_length]
grad_out = torch.randn(batch_size, device='cuda', dtype=torch.float)
with torch.backends.cudnn.flags(enabled=False):
    loss_native = torch.nn.functional.ctc_loss(log_probs.double(), targets, input_lengths, target_lengths, reduction='none')
    grad_native = torch.autograd.grad(loss_native, log_probs, grad_out)[0].float()
torch.autograd.gradcheck(lambda x: torch.nn.functional.ctc_loss(x.log_softmax(2), targets, input_lengths, target_lengths, reduction='sum'), log_probs.double())
loss_cudnn = torch.nn.functional.ctc_loss(log_probs, targets.to('cpu', torch.int32),
                                                   input_lengths, target_lengths, reduction='none')
grad_cudnn, = torch.autograd.grad(loss_cudnn, log_probs, grad_out)
print((grad_cudnn-grad_native).max())
```

t-vi committed:
onlyCUDA attribute

t-vi commented:
Just a quick note: The obvious alternative is to just support CuDNN >= 7.6 and not hack around 7.1 compatibility issues.

soumith commented:
please disable cudnn path for CuDNN < 7.6 and let's get this patch in

t-vi committed:
Don't compile in CuDNN CTC Loss unless CuDNN is >= 7.6

t-vi committed:
don't define setEx unless CuDNN is >= 7.6

t-vi committed:
bump tests' version checks

t-vi commented:
This is still broken, apparently, e.g. `pytorch_namedtensor_linux_xenial_cuda9_cudnn7_py2_test`.

t-vi committed:
move targets to GPU if needed to be more backend-agnostic

t-vi committed:
Merge branch 'master' into cudnn_ctc_gradout

t-vi committed:
move CuDNN into a _use_cudnn_ctc_loss function in native/cudnn/LossCTC.cpp

t-vi committed:
check cuda before checking cudnn

t-vi committed:
fix signature for non-cuda

t-vi committed:
Merge branch 'master' into cudnn_ctc_gradout
